Automattic/woocommerce-payments
# Add progressive onboarding tracking events

Fixes #5478

#### Changes proposed in this Pull Request
Update onboarding flow handle exit to check `document.referrer` to avoid redirecting back to Jetpack connection page.

Add the following events:
- `wcadmin_wcpay_onboarding_flow_started` – On onboarding flow load.
- `wcadmin_wcpay_onboarding_flow_mode_selected` – After clicking on **Continue** in the mode choice step.
  - Properties:
    - **mode**: `live` or `test`.
    - **elapsed**: seconds since start of onboarding.
- `wcadmin_wcpay_onboarding_flow_step_completed` – After successfully completing a form step for the first time.
  - Properties:
    - **step**: `personal`, `business` or `store`.
    - **elapsed**: seconds since the previous step.
- `wcadmin_wcpay_onboarding_flow_redirected` – Just before being redirected to Stripe KYC.
  - Properties:
    - **is_po_eligible**: `true` or `false`.
    - **elapsed**: seconds since start of onboarding.
- `wcadmin_wcpay_onboarding_flow_eligibility_modal_closed` – After closing the modal by any action.
  - Properties:
    - **action**: `dismiss`, `setup_deposits` or `enable_payments_only`.
- `wcadmin_wcpay_onboarding_flow_exited` – When the flow is abandoned by clicking on exit (`X`)
  - Properties:
    - **step**: `mode`, `personal`, `business`, `store` or `loading`.
    - **errored**: list of touched form fields with error.
    - **elapsed**: seconds since start of onboarding.
- `wcadmin_wcpay_onboarding_flow_hidden` –  Whe the flow gets hidden (more on this below).
  - Properties:
    - **step**: `mode`, `personal`, `business`, `store` or `loading`.
    - **errored**: list of touched form fields with error.
    - **elapsed**: seconds since start of onboarding.

The current implementation for the `hidden` event rely on `visibilitychange` event, [the only known way to properly handle it on any device](https://developer.chrome.com/blog/page-lifecycle-api/). It triggers when the user switch a tab, close the browser, go home on their phone, etc.

#### Testing instructions
To check that Tracks events are properly registered, you can use one of the following methods:
- Run `localStorage.setItem( 'debug', 'wc-admin:tracks' );` on your site browser console. And look for `verbose` logs.
- Install [Tracks Vigilante](https://github.com/Automattic/tracks-chrome-extension) browser extension and use its log of events.
- Go to [Tracks Live view](https://mc.a8c.com/tracks/live/), filter by any of the allowed methods, and wait patiently (~10min) for events to be registered there.

Go through the onboarding process ensuring every new track event gets registered accordingly. Apart from the description you can see [a table with screenshots here](https://docs.google.com/document/d/1XXPpXJtYPKkmOQxD43el8Mk6JTuw1IGG1SskKOvgYHE/edit).

-------------------


**Post merge**
